### PR TITLE
feat: [CDS-58244]: restrict readonly secret manager for terraform/terragrunt steps

### PR DESCRIPTION
**Pull request by sonaligoyal-harness**
### Summary

Show warning message for secret vault connector and disabled apply button if it is read only connector

#### Screenshots

SingleSelect:

Screenshot 2023-04-12 at 3 21 29 PM

MultiSelect:

https://user-images.githubusercontent.com/98508399/234246148-7fc2cc79-5a09-4034-82aa-da0427c35031.mov

#### PR Checklist

<details>
<summary>Please check if your PR fulfils the following requirements</summary>

- [ ] Tests for the changes have been added. Ideally, include a test that fails without this PR but passes with it.
- [ ] Docs have been [added/updated](https://harness.atlassian.net/jira/software/c/projects/DOC/boards/40).
</details>

<details>
<summary>Use the following comments to re-trigger PR Checks</summary>

- Jest: `retrigger jest`
- Prettier: `retrigger prettier`
- Type Check: `retrigger typecheck`
- ESLint: `retrigger eslint`
- Sonar: `retrigger sonar`
- Standards: `retrigger standards`
- Build: `retrigger build`
- Title Check: `retrigger titlecheck`
- Feature Name Check: `trigger featurenamecheck`
- Coverage: `retrigger coverage`
- Rebase: `trigger rebase`
- Cypress Rest: `retrigger cypress-rest`
- Cypress CD: `retrigger cypress-cd`
- Cypress Pipeline: `retrigger cypress-pipeline`
- Cypress CV: `retrigger cypress-cv`
- Fix Prettier: `fix prettier`
</details>

### CODE REVIEW

**Review by rameshx — CHANGES_REQUESTED** [review]
Components like `EntityReference`, `ReferenceSelect`, etc. should not take a prop like `isInfraProvisionerSecretManager`. The prop should be generic.

File: src/modules/35-connectors/components/ConnectorReferenceField/FormMultiTypeConnectorField.tsx [hunk 1]
```diff
@@ -496,6 +499,7 @@ export const MultiTypeConnectorField = (props: MultiTypeConnectorFieldProps): Re
         value={selectedValue}
         multitypeInputValue={multitypeInputValue}
         {...multiTypeProps}
+        isInfraProvisionerSecretManager={isInfraProvisionerSecretManager}
```
Comment (chetanvekariya-harness) [inline src/modules/35-connectors/components/ConnectorReferenceField/FormMultiTypeConnectorField.tsx:502]: Common components should not be outside context aware. This prop is specific. We can pass common/generic prop that takes care of readonly part inside this component. That way, consumers will be free to implement their own logic for readonly state. 
Please rename prop name to something generic.

**Review by chetanvekariya-harness — CHANGES_REQUESTED** [review]

File: src/modules/10-common/components/EntityReference/EntityReference.tsx [hunk 2]
```diff
@@ -225,6 +228,11 @@ export function EntityReference<T extends ScopedObjectDTO>(props: EntityReferenc
   const firstUpdate = useRef(true)
   const controllerRef = useRef<AbortController>()
 
+  const isReadonlySecretManager =
+    CDS_NOT_ALLOW_READ_ONLY_SECRET_MANAGER_TERRAFORM_TERRAGRUNT_PLAN &&
```
Comment (chetanvekariya-harness) [inline src/modules/10-common/components/EntityReference/EntityReference.tsx:232]: This FF logic also should be moved outside.

**Review by chetanvekariya-harness — CHANGES_REQUESTED** [review]

**Review by satyam-harness — APPROVED** [review]

File: src/modules/75-cd/components/PipelineSteps/TerraformPlan/TerraformPlan.tsx [hunk 3]
```diff
@@ -603,6 +607,23 @@ function TerraformPlanWidget(
...
+                    }
+                    renderRecordDisabledWarning={val => {
+                      return val ? (
+                        <Text
+                          icon="warning-icon"
+                          iconProps={{ size: 18, color: Color.RED_800, padding: { right: 'xsmall' } }}
+                          className={css.warningMessage}
+                        >
+                          {getString('common.readOnlyConnectorWarning')}
+                        </Text>
```
Comment (rameshx) [inline src/modules/75-cd/components/PipelineSteps/TerraformPlan/TerraformPlan.tsx:622]: can be moved to a component to reuse

**Review by rameshx — APPROVED** [review]

File: src/modules/75-cd/components/PipelineSteps/TerraformPlan/TerraformPlan.tsx (hunk 3 above)
Comment (sonaligoyal-harness) [reply src/modules/75-cd/components/PipelineSteps/TerraformPlan/TerraformPlan.tsx:622]: It was minor warning message. So didn't think of making a common one.
Should we make it ?

File: src/modules/10-common/components/ReferenceSelect/ReferenceSelect.tsx [hunk 4]
```diff
@@ -76,6 +76,8 @@ export interface ReferenceSelectProps<T extends MinimalObject>
   isOnlyFixedtype?: boolean
   selectedReferences?: string[] | Item[]
   onMultiSelectChange?: (records: ScopeAndIdentifier[]) => void
+  isRecordDisabled?: (item: any) => boolean
```
Comment (chetanvekariya-harness) [inline src/modules/10-common/components/ReferenceSelect/ReferenceSelect.tsx:79]: any?

File: src/modules/35-connectors/components/ConnectorReferenceField/ConnectorReferenceField.tsx [hunk 5]
```diff
@@ -248,6 +250,8 @@ interface GetReferenceFieldMethodProps extends ConnectorReferenceFieldProps {
   type: ConnectorInfoDTO['type'] | ConnectorInfoDTO['type'][]
   setPagedConnectorData: (data: ResponsePageConnectorResponse) => void
   version?: string
+  isRecordDisabled?: (item: any) => boolean
```
Comment (chetanvekariya-harness) [inline src/modules/35-connectors/components/ConnectorReferenceField/ConnectorReferenceField.tsx:253]: any?

File: src/modules/35-connectors/components/ConnectorReferenceField/ConnectorReferenceField.tsx [hunk 6]
```diff
@@ -129,6 +129,8 @@ export interface ConnectorReferenceFieldProps extends Omit<IFormGroupProps, 'lab
   isMultiSelect?: boolean
   selectedConnectors?: ConnectorSelectedValue[] | string[]
   onMultiSelectChange?: (arg: ScopeAndIdentifier[]) => void
+  isRecordDisabled?: (val?: any) => boolean
```
Comment (chetanvekariya-harness) [inline src/modules/35-connectors/components/ConnectorReferenceField/ConnectorReferenceField.tsx:132]: any?

File: src/modules/35-connectors/components/ConnectorReferenceField/FormMultiTypeConnectorField.tsx [hunk 7]
```diff
@@ -95,6 +95,8 @@ export interface MultiTypeConnectorFieldProps extends Omit<ConnectorReferenceFie
     templateValue: string | SelectOption | undefined
   }
   version?: string
+  isRecordDisabled?: (item: any) => boolean
```
Comment (chetanvekariya-harness) [inline src/modules/35-connectors/components/ConnectorReferenceField/FormMultiTypeConnectorField.tsx:98]: any?

File: src/modules/75-cd/components/PipelineSteps/TerraformPlan/TerraformPlan.tsx (hunk 3 above)
Comment (rameshx) [reply src/modules/75-cd/components/PipelineSteps/TerraformPlan/TerraformPlan.tsx:622]: Not necessary. Ignore 🙂

File: src/modules/75-cd/components/PipelineSteps/TerraformPlan/TerraformPlan.tsx [hunk 8]
```diff
@@ -603,6 +607,23 @@ function TerraformPlanWidget(
                     multiTypeProps={{ expressions, allowableTypes }}
                     gitScope={{ repo: repoIdentifier || '', branch, getDefaultFromOtherRepo: true }}
                     disabled={readonly}
+                    isRecordDisabled={selectedRecord =>
+                      (selectedRecord as any)?.spec?.readOnly &&
+                      CDS_NOT_ALLOW_READ_ONLY_SECRET_MANAGER_TERRAFORM_TERRAGRUNT_PLAN
+                    }
+                    renderRecordDisabledWarning={val => {
```
Comment (chetanvekariya-harness) [inline src/modules/75-cd/components/PipelineSteps/TerraformPlan/TerraformPlan.tsx:614]: We should not take any argument here. We can directly render component,

File: src/modules/35-connectors/components/ConnectorReferenceField/ConnectorReferenceField.tsx (hunk 5 above)
Comment (sonaligoyal-harness) [reply src/modules/35-connectors/components/ConnectorReferenceField/ConnectorReferenceField.tsx:253]: We have it is as any only for record in index file also.

File: src/modules/75-cd/components/PipelineSteps/TerraformPlan/TerraformPlan.tsx (hunk 8 above)
Comment (sonaligoyal-harness) [reply src/modules/75-cd/components/PipelineSteps/TerraformPlan/TerraformPlan.tsx:614]: How can we distinguish for not showing if no connector is readonly?

Comment (chetanvekariya-harness) [reply src/modules/75-cd/components/PipelineSteps/TerraformPlan/TerraformPlan.tsx:614]: If it is readonly then only call this function. Is it no possible?

File: src/modules/10-common/components/EntityReference/EntityReference.tsx [hunk 9]
```diff
@@ -420,7 +436,7 @@ export function EntityReference<T extends ScopedObjectDTO>(props: EntityReferenc
           {renderTab(true, TAB_ID.ACCOUNT, 'account', 'layers', selectedAccount?.accountName)}
         </Tabs>
       </div>
-
+      {renderRecordDisabledWarning?.(isMultiSelect ? isMultiSelectRecordDisabled : isRecordDisabled?.(selectedRecord))}
```
Comment (chetanvekariya-harness) [inline src/modules/10-common/components/EntityReference/EntityReference.tsx:439]: Call this render function only when flag is true. Is it possible to do?

File: src/modules/75-cd/components/PipelineSteps/TerraformPlan/TerraformPlan.tsx (hunk 8 above)
Comment (sonaligoyal-harness) [reply src/modules/75-cd/components/PipelineSteps/TerraformPlan/TerraformPlan.tsx:614]: yes, updating it

File: src/modules/10-common/components/EntityReference/EntityReference.tsx (hunk 9 above)
Comment (sonaligoyal-harness) [reply src/modules/10-common/components/EntityReference/EntityReference.tsx]: done

**Review by chetanvekariya-harness — APPROVED** [review]

File: src/modules/10-common/components/EntityReference/EntityReference.tsx [hunk 10]
```diff
@@ -421,6 +437,9 @@ export function EntityReference<T extends ScopedObjectDTO>(props: EntityReferenc
         </Tabs>
       </div>
 
+      {((isMultiSelect && isMultiSelectRecordDisabled) || isRecordDisabled?.(selectedRecord)) &&
+        renderRecordDisabledWarning}
```
[the remark this answers is not in the document]
Comment (sonaligoyal-harness) [reply src/modules/10-common/components/EntityReference/EntityReference.tsx:441]: sure

**Review by rameshx — DISMISSED** [review]

**Review by satyam-harness — APPROVED** [review]

**Review by sunnykesh-harness — APPROVED** [review]

**Pull request merged**